compiler/rustc_metadata/src/rmeta/decoder/cstore_impl.rs

```diff
@@ -44,13 +44,15 @@ macro_rules! provide {
                 let ($def_id, $other) = def_id_arg.into_args();
                 assert!(!$def_id.is_local());
 
-                let $cdata = CStore::from_tcx($tcx).get_crate_data($def_id.krate);
-
-                if $tcx.dep_graph.is_fully_enabled() {
-                    let crate_dep_node_index = $cdata.get_crate_dep_node_index($tcx);
-                    $tcx.dep_graph.read_index(crate_dep_node_index);
+                // External query providers call `crate_hash` in order to register a dependency
+                // on the crate metadata. The exception is `crate_hash` itself, which obviously
+                // doesn't need to do this (and can't, as it would cause a query cycle).
+                if stringify!($name) != "crate_hash" && $tcx.dep_graph.is_fully_enabled() {
```

Comment:
I should have suggested this in the first place:
`rustc_middle::dep_graph::DepKind::$name != rustc_middle::dep_graph::DepKind::crate_hash`.
This will err at compile time in addition to the comment.